api/citations/utils.py

```diff
@@ -75,4 +110,148 @@ def render_citation(node, style='apa'):
     elif cit.count(title) == 0:
         cit = clean_up_common_errors(cit)
 
+    if isinstance(node, PreprintService):
+        cit_node = node.node
+    else:
+        cit_node = node
+
+    if style == 'apa':
+        cit = apa_reformat(cit_node, cit)
+    if style == 'chicago-author-date':
+        cit = chicago_reformat(cit_node, cit)
+    if style == 'modern-language-association':
+        cit = mla_reformat(cit_node, cit)
+
+    return cit
+
+
+def apa_reformat(node, cit):
+    new_csl = cit.split('(')
+    contributors_list = [x for x in node.contributors if node.get_visible(x)]
+
+    # throw error if there is no visible contributor
+    if len(contributors_list) == 0:
+        raise HTTPError(http.BAD_REQUEST)
+    # handle only one contributor
+    elif len(contributors_list) == 1:
+        name = process_name(node, contributors_list[0])
+        new_apa = apa_name(name)
+    # handle more than one contributor  but less than 8 contributors
+    elif len(contributors_list) in range(1, 8):
+        name_list = [apa_name(process_name(node, x)) for x in contributors_list[:-1]]
+        new_apa = ' '.join(name_list)
+        last_one = apa_name(process_name(node, contributors_list[-1]))
+        new_apa += ' & ' + last_one
+    # handle 8 or more contributors
+    else:
+        name_list = [apa_name(process_name(node, x)) for x in contributors_list[:5]]
+        new_apa = ' '.join(name_list) + '... ' + apa_name(process_name(node, contributors_list[-1]))
+
+    cit = new_apa.rstrip(', ') + ' '
+    for x in new_csl[1:]:
+        cit += '(' + x
+    return cit
+
+
+def apa_name(name):
+    apa = ''
+    if name['family_name']:
+        apa += name['family_name'] + ','
+    if name['given_name']:
+        apa += ' ' + name['given_name'][0] + '.'
+        if name['middle_names']:
+            apa += ' ' + name['middle_names'][0] + '.'
+        apa += ','
+    if name['suffix']:
+        apa += ' ' + name['suffix'] + ','
+    return apa
+
+
+def mla_reformat(node, cit):
+    contributors_list = [x for x in node.contributors if node.get_visible(x)]
+    retrive_from = cit.split('Open')[-1]
+    # throw error if there is no visible contributor
+    if len(contributors_list) == 0:
+        raise HTTPError(http.BAD_REQUEST)
+    # handle only one contributor
+    elif len(contributors_list) == 1:
```

Comment:
Is L177-189 identical to L207-219? If so, you might as well make that a function.